src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs

```diff
@@ -0,0 +1,139 @@
+﻿
+
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.MoveToNamespace
+{
+    internal abstract class AbstractMoveToNamespaceService : ILanguageService
+    {
+        internal abstract Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(Document document, TextSpan span, CancellationToken cancellationToken);
+        internal abstract Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(Document document, int position, CancellationToken cancellationToken);
+        public abstract Task<MoveToNamespaceResult> MoveToNamespaceAsync(MoveToNamespaceAnalysisResult analysisResult, string targetNamespace, CancellationToken cancellationToken);
+        public abstract Task<MoveToNamespaceOptionsResult> GetOptionsAsync(Document document, string defaultNamespace, CancellationToken cancellationToken);
+    }
+
+    internal abstract class AbstractMoveToNamespaceService<TCompilationSyntax, TNamespaceDeclarationSyntax, TNamedTypeDeclarationSyntax>
+        : AbstractMoveToNamespaceService
+    {
+        private IMoveToNamespaceOptionsService _moveToNamespaceOptionsService;
+
+        protected abstract string GetNamespaceName(TNamespaceDeclarationSyntax syntax);
+        protected abstract string GetNamespaceName(TNamedTypeDeclarationSyntax syntax);
+
+        public AbstractMoveToNamespaceService(IMoveToNamespaceOptionsService moveToNamespaceOptionsService)
+        {
+            _moveToNamespaceOptionsService = moveToNamespaceOptionsService;
+        }
+
+        internal override async Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var typeAnalysisResult = await AnalyzeTypeAtPositionAsync(document, span.Start, cancellationToken).ConfigureAwait(false);
+
+            if (typeAnalysisResult.CanPerform)
+            {
+                return ImmutableArray.Create(AbstractMoveToNamespaceCodeAction.Generate(this, typeAnalysisResult));
+            }
+
+            return ImmutableArray<AbstractMoveToNamespaceCodeAction>.Empty;
+        }
+
+        internal override async Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(
+            Document document,
+            int position,
+            CancellationToken cancellationToken)
+        {
+#if DEBUG // TODO: remove once the feature is done
```

Comment:
Should we be ignoring this bit for now?

Reply:
Yes, there will be one more review where you can pick this one apart.